### PR TITLE
[DEV-257] Mint as raft holder

### PR DESCRIPTION
**Pull request by grp06**

### CODE REVIEW

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
The submodule for ERC4973 should not be checked in here. Did you intend to include this?

File: src/Badges.t.sol [hunk 1]
```diff
@@ -1,63 +1,63 @@
 // SPDX-License-Identifier: CC0-1.0
-pragma solidity ^0.8.15;
+// pragma solidity ^0.8.15;
```
Comment (rahulrumalla) [inline src/Badges.t.sol:2]: Either let's update these tests or delete this file. My recommendation would be to unit test this in forge and integration test it in hardhat.

File: test/badges.test.ts [hunk 2]
```diff
@@ -8,129 +8,153 @@ import { splitSignature, _TypedDataEncoder } from 'ethers/lib/utils'
...
-                ],
-            },
-            value: {
-                from: issuer.address,
-                to: claimant.address,
-                tokenURI,
-            },
-        }
-        return { badges, badgesContract, owner, issuer, claimant, badActor, typedData }
+  const name = 'Otter'
```
Comment (rahulrumalla) [inline test/badges.test.ts:11]: qq, why aren't we reading these from environment variables?

File: scripts/deploy_and_verify_raft.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
Comment (rahulrumalla) [inline scripts/deploy_and_verify_raft.sh:1]: hey, wasn't this already merged from another story? 🤔

File: scripts/deploy_and_verify_raft.sh [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+# $1: the name of the .env file to use
+source $1
+forge create src/Raft.sol:Raft --constructor-args $OWNER_ADDRESS $NAME $SYMBOL --private-key $PRIVATE_KEY --verify --rpc-url $ETH_RPC_URL --etherscan-api-key $ETHERSCAN_API_KEY
```
Comment (rahulrumalla) [inline scripts/deploy_and_verify_raft.sh:4]: Similar to convention followed in the other script, I'd recommend using $RAFT_NAME $RAFT_SYMBOL $RAFT_OWNER_ADDRESS

Also maybe it's more prescriptive to say $DEPLOYER_PRIVATE_KEY

File: src/Badges.sol [hunk 5]
```diff
@@ -17,4 +21,44 @@ contract Badges is ERC4973Permit {
...
+    uint256 raftTokenId = _specToRaft[specUri];
+    // only registered specs can be used for minting
+    require(raftTokenId != 0, "mintAuthorizedBadge: spec is not registered");
+    // if we use this.take() it will pass in this contract's address as msg.sender
+    // so we use delegatecall to make sure we use the caller's address as msg.sender
+    (bool success, bytes memory data) = address(this).delegatecall(
+      abi.encodeWithSignature("take(address,string,bytes)", from, specUri, signature)
+    );
+    if (!success) {
+      revert("mintAuthorizedBadge: badge minting failed");
```
Comment (rahulrumalla) [inline src/Badges.sol:46]: Why not use the following?

```
revert(success, "mintAuthorizedBadge: badge minting failed");
```

File: src/Badges.sol [hunk 6]
```diff
@@ -17,4 +21,44 @@ contract Badges is ERC4973Permit {
...
+    }
+    
+    uint256 tokenId = abi.decode(data, (uint256));
+    _badgeToRaft[tokenId] = raftTokenId;
+    emit BadgeMinted(from, tokenId);
+    return tokenId;
+  }
+
+  function createSpecAsRaftOwner(string memory specUri, uint256 raftTokenId) external {
+    require(_specToRaft[specUri] == 0, "spec already registered");
```
Comment (rahulrumalla) [inline src/Badges.sol:56]:
```suggestion
    require(_specToRaft[specUri] == 0, "createSpecAsRaftOwner: spec already registered");
```

File: test/badges.test.ts [hunk 7]
```diff
@@ -8,129 +8,153 @@ import { splitSignature, _TypedDataEncoder } from 'ethers/lib/utils'
...
+    expect(deployedContractName).to.equal(name)
+    expect(deployedSymbolName).to.equal(symbol)
+
+    const provider = waffle.provider
+    const network = await provider.getNetwork()
+    const deployedToChainId = network.chainId
+    expect(deployedToChainId).to.equal(chainId)
+  })
+
+  it('should match off-chain hash to on-chain hash', async () => {
```
Comment (rahulrumalla) [inline test/badges.test.ts:79]: Hmm.. negative case coverage is missing I think. Could we add?

File: test/badges.test.ts [hunk 8]
```diff
@@ -8,129 +8,153 @@ import { splitSignature, _TypedDataEncoder } from 'ethers/lib/utils'
...
+      typedData.value.passive,
+      typedData.value.tokenURI
+    )
+
+    expect(offChainHash).to.equal(onChainHash)
+  })
+
+  it('should fail to mintWithAuthorizedBadge when using incorrect issuer address', async () => {
+    const { badgesContract, typedData, issuer, claimant } = await loadFixture(deployContractFixture)
+    const signature = await issuer._signTypedData(typedData.domain, typedData.types, typedData.value)
```
Comment (rahulrumalla) [inline test/badges.test.ts:94]: I think you could expand this by including another test case i.e., using incorrect issuer address and also the signature from the incorrect address

File: test/badges.test.ts [hunk 9]
```diff
@@ -8,129 +8,153 @@ import { splitSignature, _TypedDataEncoder } from 'ethers/lib/utils'
...
+
+  it('should fail to mintWithAuthorizedBadge when using incorrect token uri', async () => {
+    const { badgesContract, typedData, issuer, claimant } = await loadFixture(deployContractFixture)
+    const signature = await issuer._signTypedData(typedData.domain, typedData.types, typedData.value)
+    const { compact } = splitSignature(signature)
+
+    await expect(
+      badgesContract
+        .connect(claimant)
+        .mintAuthorizedBadge(typedData.value.passive, 'https://some-incorrect-uri.com', compact)
```
Comment (rahulrumalla) [inline test/badges.test.ts:111]: Can we use the uri as a const than hard-code, please? 🙏🏽

File: test/badges.test.ts [hunk 10]
```diff
@@ -8,129 +8,153 @@ import { splitSignature, _TypedDataEncoder } from 'ethers/lib/utils'
...
+
+    await expect(
+      badgesContract
+        .connect(claimant)
+        .mintAuthorizedBadge(typedData.value.passive, typedData.value.tokenURI, sigAsBytes)
+    ).to.be.revertedWith('mintAuthorizedBadge: badge minting failed')
+  })
+
+  it('should fail to mintWithAuthorizedBadge when using unauthorized claimant', async () => {
+    const { badgesContract, typedData, issuer, badActor } = await loadFixture(deployContractFixture)
```
Comment (rahulrumalla) [inline test/badges.test.ts:128]: Maybe it's better to rename `badActor` to `claimant2` or `randomSigner` ? Could be confusing to infer what a badActor could mean in general and also within this context..

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
Hey, I've gone through all the changes and left notes. Please lmk if you need to discuss / chat. 

Another thing I noticed is that this seems to miss coverage for "spec creation" in general?

File: src/Badges.sol (hunk 5 above)
Comment (grp06) [reply src/Badges.sol:46]: Revert can't take a condition like `require` ... It just takes a string. So you have to perform a check first

File: test/badges.test.ts (hunk 7 above)
Comment (grp06) [reply test/badges.test.ts:79]: can you elaborate on this one? I don't follow

Comment (rahulrumalla) [reply test/badges.test.ts:79]: actually, scratch that. My bad. It won't make sense to have a negative case coverage here.

File: test/badges.test.ts (hunk 2 above)
Comment (rahulrumalla) [reply test/badges.test.ts:11]: Discussed in DM. Resolving this.

**Comment by grp06** [comment]
> The submodule for ERC4973 should not be checked in here. Did you intend to include this?

Didn't mean to. I'll remove it

File: src/Badges.t.sol (hunk 1 above)
Comment (grp06) [reply src/Badges.t.sol:2]: Agreed. I'll take it out for now and make a ticket in linear to write proper unit tests

**Comment by grp06** [comment]
> Hey, I've gone through all the changes and left notes. Please lmk if you need to discuss / chat.
> 
> Another thing I noticed is that this seems to miss coverage for "spec creation" in general?

added

**Pull request merged**